executor/aggfuncs/func_group_concat.go

```diff
@@ -93,27 +97,43 @@ func (e *groupConcat) UpdatePartialResult(sctx sessionctx.Context, rowsInGroup [
 			if isNull {
 				continue
 			}
-			isWriteSep = true
-			if p.buffer == nil {
-				p.buffer = &bytes.Buffer{}
-			}
-			p.buffer.WriteString(v)
+			isAllNull = false
+			p.valsBuf.WriteString(v)
+		}
+		if isAllNull {
+			continue
 		}
-		if isWriteSep {
+		if p.buffer == nil {
+			p.buffer = &bytes.Buffer{}
+		} else {
 			p.buffer.WriteString(e.sep)
 		}
+		p.buffer.WriteString(p.valsBuf.String())
 	}
-	if p.buffer != nil {
-		p.buffer.Truncate(p.buffer.Len() - len(e.sep))
```

Comment:
I think the bug is caused by this statement.

We can change the strategy to write the separator to this to avoid a double string copy:

```
for every input record:
	1. if all the fields are NULL, skip this record, else goto step 2.
	2. check whether the buffer is empty, if not, write the separator.
	3. write the non-NULL fields into the buffer.
```

Reply:
I think you can file another to solve this bug, as you can see, this PR is very huge now.